### PR TITLE
[14.0][IMP] stock_request: Add option to use stock available

### PR DESCRIPTION
**Pull request by victoralmau**
Add option to use stock available

Please @pedrobaeza and @chienandalu can you review it?

@Tecnativa TT43413

### CODE REVIEW

File: stock_request/models/res_company.py [hunk 1]
```diff
@@ -12,3 +12,6 @@ class ResCompany(models.Model):
     stock_request_allow_virtual_loc = fields.Boolean(
         string="Allow Virtual locations on Stock Requests"
     )
+    stock_request_check_available_stock_first = fields.Boolean(
```
Comment (pedrobaeza) [inline stock_request/models/res_company.py]: Simplify variable name at maximum, as stock is already mentioned:

```suggestion
    stock_request_check_available_first = fields.Boolean(
```

File: stock_request/models/stock_request.py [hunk 2]
```diff
@@ -258,8 +258,9 @@ def check_order_picking_policy(self):
             raise ValidationError(_("The picking policy must be equal to the order"))
 
     def _action_confirm(self):
+        self._action_use_stock_available()
```
Comment (pedrobaeza) [inline stock_request/models/stock_request.py]: Call this method only for the SR that have it checked.

File: stock_request/models/stock_request.py [hunk 3]
```diff
@@ -340,6 +341,49 @@ def _prepare_procurement_values(self, group_id=False):
...
+        move._action_done()
+        return move
+
+    def _action_use_stock_available(self):
+        allocation_model = self.env["stock.request.allocation"]
+        stock_move_model = self.env["stock.move"].sudo()
+        for request in self.filtered(
+            lambda x: x.company_id.stock_request_check_available_stock_first
+        ):
+            if request.product_id.sudo().qty_available >= request.product_uom_qty:
```
Comment (pedrobaeza) [inline stock_request/models/stock_request.py]: You are not passing location as context for restricting the available quantity. I also understand that you need to take into account other SR allocations for reducing this quantity.

File: stock_request/models/stock_request.py [hunk 4]
```diff
@@ -340,6 +341,49 @@ def _prepare_procurement_values(self, group_id=False):
     def _skip_procurement(self):
         return self.state != "draft" or self.product_id.type not in ("consu", "product")
 
+    def _prepare_stock_move(self):
+        return {
```
Comment (pedrobaeza) [inline stock_request/models/stock_request.py:340]: This phantom move will be garbage and will corrupt statistics like forecasting, so we should try to get another way. And if not, at least `location_id` and `location_dest_id` should be both `location.id` in my understanding.

I know that allocations require a move to be linked, but I think we can relax such requirement. The only place where it's used is for checking its state for putting the opened qty.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by victoralmau** [comment]
I have made some changes and added more use cases.

I believe that stock move is necessary, and in some cases several, for example when you have several child locations (the stock you need may be entirely in one child location, or you may need to use several locations).

**Comment by victoralmau** [comment]
Ping @pedrobaeza  and @chienandalu

**Comment by victoralmau** [comment]
All changes are done, can you check it again?

File: stock_request/models/stock_request.py [hunk 5]
```diff
@@ -362,6 +394,14 @@ def _action_launch_procurement_rule(self):
             if float_compare(qty, request.product_qty, precision_digits=precision) >= 0:
                 continue
 
+            # If stock is available we use it and we do not execute rule
+            if request.company_id.stock_request_check_available_first:
+                quants = quant_model._gather(request.product_id, request.location_id)
+                qty_available = sum(quants.mapped("available_quantity"))
+                if qty_available >= request.product_uom_qty:
```
Comment (rousseldenis) [inline stock_request/models/stock_request.py]: Please use float_compare

File: stock_request/models/stock_request.py [hunk 6]
```diff
@@ -259,7 +259,7 @@ def check_order_picking_policy(self):
 
     def _action_confirm(self):
         self._action_launch_procurement_rule()
-        self.write({"state": "open"})
+        self.filtered(lambda x: x.state != "done").write({"state": "open"})
```
Comment (pedrobaeza) [inline stock_request/models/stock_request.py:262]: Isn't this a computed field now?

Comment (victoralmau) [reply stock_request/models/stock_request.py:262]: Not yet, it is related to https://github.com/OCA/stock-logistics-warehouse/pull/1822

File: stock_request/models/stock_request.py [hunk 7]
```diff
@@ -362,6 +394,21 @@ def _action_launch_procurement_rule(self):
             if float_compare(qty, request.product_qty, precision_digits=precision) >= 0:
                 continue
 
+            # If stock is available we use it and we do not execute rule
+            if request.company_id.stock_request_check_available_first:
+                quants = quant_model._gather(request.product_id, request.location_id)
```
Comment (pedrobaeza) [inline stock_request/models/stock_request.py]: This is still not valid, as you have to make as many moves as possible locations. That's why I told you about first getting qty_available from product with the proper context, and then do the moves with `_gather`.

Comment (victoralmau) [reply stock_request/models/stock_request.py]: Changes done to take into account quants and create the necessary moves.

**Review by pedrobaeza — APPROVED** [review]

**Review by carlosdauden — APPROVED** [review]

**Comment by carlosdauden** [comment]
Please rebase

**Comment by victoralmau** [comment]
Rebase done.

**Review by chienandalu — APPROVED** [review]

**Pull request merged**